### PR TITLE
Added Unitsml::Formula to Math conversion methods

### PR DESCRIPTION
**Pull request by suleman-uzair**
Added **Unitsml** to **MathML**, **Latex** conversion methods.

### CODE REVIEW

File: lib/unitsml/utility.rb [hunk 1]
```diff
@@ -152,10 +151,9 @@ def combine_prefixes(p1, p2)
         p1
       end
 
-      def unit(units, _orig_text, norm_text, name)
-        dimid = dim_id(units2dimensions(units))
+      def unit(units, dims, _orig_text, norm_text, name)
         <<~XML
-          <Unit xmlns="#{UNITSML_NS}" xml:id="#{unit_id(norm_text)}"#{dimid ? " dimensionURL=\"##{dimid}\"" : ""}>
+          <Unit xmlns="#{UNITSML_NS}" xml:id="#{unit_id(norm_text)}"#{dims ? " dimensionURL=\"##{dim_id(dims)}\"" : ""}>
```
Comment (ronaldtse) [inline lib/unitsml/utility.rb:156]: @suleman-uzair can we not have handcrafted XML?

File: lib/unitsml/utility.rb [hunk 2]
```diff
@@ -182,17 +180,45 @@ def unitsystem(units)
...
       end
 
+      def dimension(norm_text)
+        return unless Unitsdb.units.dig(norm_text, :fields, "dimension_url")
+
+        dim_id = Unitsdb.units.dig(norm_text, :fields, "dimension_url").sub("#", '')
+        <<~XML
+          <Dimension xmlns='#{UNITSML_NS}' xml:id="#{dim_id}">
+            #{dimid2dimensions(dim_id)&.map { |u| dimension1(u) }&.join("\n  ")}
+          </Dimension>
```
Comment (ronaldtse) [inline lib/unitsml/utility.rb:190]: No handcrafted XML please.

File: lib/unitsml/utility.rb [hunk 3]
```diff
@@ -182,17 +180,45 @@ def unitsystem(units)
...
+        XML
+      end
+
+      def dimension1(dim)
+        %(<#{dim[:dimension]} symbol="#{dim[:symbol]}" powerNumerator="#{float_to_display(dim[:exponent])}"/>)
```
Comment (ronaldtse) [inline lib/unitsml/utility.rb:195]: No handcrafted XML please.

File: lib/unitsml/utility.rb [hunk 4]
```diff
@@ -182,17 +180,45 @@ def unitsystem(units)
...
-              <PrefixSymbol type="ASCII">#{p.to_asciimath}</PrefixSymbol>
-              <PrefixSymbol type="unicode">#{p.to_unicode}</PrefixSymbol>
-              <PrefixSymbol type="LaTeX">#{p.to_latex}</PrefixSymbol>
-              <PrefixSymbol type="HTML">#{p.to_html}</PrefixSymbol>
+            <Prefix xmlns='#{UNITSML_NS}' prefixBase='#{p&.base}' prefixPower='#{p&.power}' xml:id='#{p&.id}'>
+              <PrefixName xml:lang="en">#{p&.name}</PrefixName>
+              <PrefixSymbol type="ASCII">#{p&.to_asciimath}</PrefixSymbol>
+              <PrefixSymbol type="unicode">#{p&.to_unicode}</PrefixSymbol>
+              <PrefixSymbol type="LaTeX">#{p&.to_latex}</PrefixSymbol>
+              <PrefixSymbol type="HTML">#{p&.to_html}</PrefixSymbol>
```
Comment (ronaldtse) [inline lib/unitsml/utility.rb:221]: No handcrafted XML please.

File: lib/unitsml/utility.rb [hunk 5]
```diff
@@ -221,6 +247,39 @@ def unit_id(text)
...
         "U_#{unit_hash ? unit_hash[:id]&.gsub(/'/, '_') : norm_text&.gsub(/\*/, '.')&.gsub(/\^/, '')}"
       end
+
+      def dimension_components(dims)
+        return if dims.nil? || dims.empty?
+
+        <<~XML
+          <Dimension xmlns='#{UNITSML_NS}' xml:id="#{dim_id(dims)}">
+            #{dims.map { |u| dimension1(u) }.join("\n  ")}
+          </Dimension>
```
Comment (ronaldtse) [inline lib/unitsml/utility.rb:257]: No handcrafted XML please.

File: lib/unitsml/utility.rb [hunk 6]
```diff
@@ -221,6 +247,39 @@ def unit_id(text)
...
+        return unless units[normtext] && units.dig(normtext, :fields, "quantity_reference").size == 1 ||
+          Unitsdb.quantities[quantity]
+
+        id = quantity || units.dig(normtext, :fields, "quantity_reference")&.first&.dig("url")
+        units.dig(normtext, :fields, "dimension_url") and
+          dim = %( dimensionURL="##{units.dig(normtext, :fields, "dimension_url")}")
+        <<~XML
+          <Quantity xmlns='#{UNITSML_NS}' xml:id="#{id}"#{dim} quantityType="base">
+            #{quantityname(id.sub('#', ''))}
+          </Quantity>
```
Comment (ronaldtse) [inline lib/unitsml/utility.rb:272]: No handcrafted XML please.

File: lib/unitsml/utility.rb [hunk 7]
```diff
@@ -221,6 +247,39 @@ def unit_id(text)
...
+        XML
+      end
+
+      def quantityname(id)
+        ret = ""
+        Unitsdb.quantities[id]&.dig("quantity_name")&.each do |q|
+          ret += %(<QuantityName xml:lang="en-US">#{q}</QuantityName>)
```
Comment (ronaldtse) [inline lib/unitsml/utility.rb:279]: No handcrafted XML please.

File: lib/unitsml/formula.rb [hunk 8]
```diff
@@ -92,9 +92,14 @@ def extract_units(formula)
...
+      all_units = extract_units(value)
+      dims = Utility.units2dimensions(all_units)
       <<~UNITSML
-        #{Utility.unit(extract_units(value), @orig_text, @norm_text, explicit_value&.dig(:name))}
-        #{Utility.prefixes(extract_units(value))}
+        #{Utility.unit(all_units, dims, @orig_text, @norm_text, explicit_value&.dig(:name))}
+        #{Utility.prefixes(all_units)}
+        #{Utility.dimension(@norm_text)}
+        #{Utility.dimension_components(dims)}
+        #{Utility.quantity(@norm_text, explicit_value&.dig(:quantity))}
```
Comment (ronaldtse) [inline lib/unitsml/formula.rb:102]: No handcrafted XML please.

**Review by ronaldtse — CHANGES_REQUESTED** [review]
No handcrafted XML please

**Comment by suleman-uzair** [comment]
As you suggested, we have implemented support for **Ox** for **XML** conversion. However, it's important to note that the conversion of **MathML** and **HTML** is different. The **MathML** and **HTML** conversion data are available in **unitsdb**, and we are currently using this data for the conversion process. There are two important points to mention:
1. We can use string interpolation for the conversion process(`already implemented`).
2. We can use the **Ox** parse method to parse the _unitsdb_ output, which includes the values of both the MathML and **HTML** keys in the `units.yaml` file. However, it's important to note that this method may raise an exception, as seen in the following example:

```
"NISTu89":
  dimension_url: "#NISTd16"
  short:
  root: false
  unit_system:
    type: "non-SI_not_acceptable"
    name: "non-SI"
  unit_name:
    - "erg per second"
  unit_symbols:
    - id: "erg*s^-1"
      ascii: "erg*s^-1"
      html: "erg/s"
      mathml: "<mrow><mi mathvariant='normal'>erg</mi><mo>/</mo><mrow><mi mathvariant='normal'>s</mi></mrow>"
      latex: \ensuremath{\mathrm{erg/s}}
      unicode: "erg·s⁻¹"
  root_units:
    enumerated_root_units:
      - unit: "erg"
        power_denominator: 1
        power_numerator: 1
      - unit: "second"
        power_denominator: 1
        power_numerator: -1
  quantity_reference:
    - name: "power"
      url: "#NISTq20"
    - name: "radiant flux"
      url: "#NISTq21"
```
`mathml: "<mrow><mi mathvariant='normal'>erg</mi><mo>/</mo><mrow><mi mathvariant='normal'>s</mi></mrow>"`
In this example, one closing _mrow tag_ is missing, which will cause an exception when using the **Ox** parse method.

**Comment by ronaldtse** [comment]
> We can use the Ox parse method to parse the unitsdb output, which includes the values of both the MathML and HTML keys in the units.yaml file.

MathML should be parsed properly via Plurimath.

We do not really need to parse HTML, but just offer the value.

e.g.
```ruby
Units.get_unit(:NISTu89).to_html
=> "erg/s"
Units.get_unit(:NISTu89).to_mathml
=> "<mrow><mi mathvariant='normal'>erg</mi><mo>/</mo><mrow><mi mathvariant='normal'>s</mi></mrow>"
Units.get_unit(:NISTu89).to_plurimath
=> Plurimath::Math
```

### Broken MathML

> mathml: `"<mrow><mi mathvariant='normal'>erg</mi><mo>/</mo><mrow><mi mathvariant='normal'>s</mi></mrow>"`
>
> In this example, one closing mrow tag is missing, which will cause an exception when using the Ox parse method.

We should fix this in UnitsDB. We should have a validation step (via rspec) that ensures all values in UnitsDB are valid.

**Comment by suleman-uzair** [comment]
> We should fix this in UnitsDB. We should have a validation step (via rspec) that ensures all values in UnitsDB are valid.

Should we add the rspec test to the UnitsDB repo for the validation of MathML?

**Comment by ronaldtse** [comment]
> Should we add the rspec test to the UnitsDB repo for the validation of MathML?

Yes please!

**Comment by ronaldtse** [comment]
@suleman-uzair is this ready to merge now? Thanks.

**Comment by suleman-uzair** [comment]
> @suleman-uzair is this ready to merge now? Thanks.

@ronaldtse , As we have made some changes to the **Unitsdb** **MathML** output, I am including **Ox** parsing functionality to generate the **MathML** in **Unitsml-Ruby**.

File: lib/unitsml.rb [hunk 9]
```diff
@@ -15,3 +15,5 @@ def self.parse(string)
     Unitsml::Parser.new(string).parse
   end
 end
+
+Ox.default_options = { encoding: "UTF-8" }
```
Comment (ronaldtse) [inline lib/unitsml.rb:19]: This sets a global option and it's dangerous -- Ox is used also in other gems. I think the `unitsml` gem needs to use its own Ox instance.

File: lib/unitsml/formula.rb [hunk 10]
```diff
@@ -22,5 +24,139 @@ def ==(object)
...
+      end
+
+      dimensions
+    end
+
+    def extract_units(formula)
+      units_arr = []
+
+      formula.each do |term|
+        if term.is_a?(Unit)
```
Comment (ronaldtse) [inline lib/unitsml/formula.rb:100]: Use a `case` `when` statement.

File: lib/unitsml/formula.rb [hunk 11]
```diff
@@ -22,5 +24,139 @@ def ==(object)
...
+      units_arr
+    end
+
+    def units
+      all_units = extract_units(value)
+      norm_text = Utility.postprocess_normtext(all_units)
+      dims = Utility.units2dimensions(extract_units(value))
+      dimension, dimension_component = unique_dimensions(dims, norm_text)
+
+      <<~UNITSML
```
Comment (ronaldtse) [inline lib/unitsml/formula.rb:118]: Use a `[].join`

File: lib/unitsml/formula.rb [hunk 12]
```diff
@@ -22,5 +24,139 @@ def ==(object)
...
+      dims_hash = Utility::Dim2D
+      values.sort do |first, second|
+        dims_hash.dig(first.dimension_name, :order) <=> dims_hash.dig(second.dimension_name, :order)
+      end
+    end
+
+    def prefixes
+      norm_text = @norm_text&.split("-")&.first
+      prefix_object = Unit.new("", prefix: Prefix.new(norm_text))
+      <<~XML
```
Comment (ronaldtse) [inline lib/unitsml/formula.rb:155]: Use `[].join`

File: lib/unitsml/unit.rb [hunk 13]
```diff
@@ -2,24 +2,112 @@
...
+    end
+
+    def unit_symbols
+      symbols = fields_hash.dig("unit_symbols")
+      symbols.find{|symbol| symbol["id"] == unit_name }
+    end
+
+    def numerator_value(mathml = true)
+      integer = power_numerator.to_s
+      if integer.match?(/-/)
```
Comment (ronaldtse) [inline lib/unitsml/unit.rb:33]: Use an early return form:
```ruby
unless integer.match?(/-/)
  return mathml ? [Ox.parse("<mn>#{integer}</mn>")] : integer
end

if mathml
...
```

**Review by ronaldtse — CHANGES_REQUESTED** [review]

**Comment by suleman-uzair** [comment]
@ronaldtse, I have completed the implementation of the changes. Please let me know if you would like me to proceed with merging.

**Pull request merged**